### ISSUE TITLE
Undo/redo don't work after the file from the previous session has been closed

### ISSUE THREAD

**Issue by AlexWayfer** [atom/atom#15100]
### Prerequisites

* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

*From title of issue:*

Undo/redo don't work after the file from the previous session has been closed.

**Note**: In Safe Mode undo/redo don't work for any closed file, not only from the previous session.

### Steps to Reproduce

1. Open existing file (or create the new one)
2. Edit this file
3. Close this file
4. Restore the last closed tab (`Ctrl+Shift+T`)
    1. Undo/redo work (not in Safe Mode)
5. Close Atom
6. Reopen Atom
    1. Undo/redo work
7. Close this file
8. Reopen this file (as in step 4)
    1. Undo/redo don't work!

**Expected behavior:** Undo/redo work in any time and combination of situations.

**Actual behavior:** Undo/redo don't work in certain combination of situations.

**Reproduces how often:** 100%

### Versions

```
$ atom --version
Atom    : 1.18.0
Electron: 1.3.15
Chrome  : 52.0.2743.82
Node    : 6.5.0

$ apm --version
apm  1.18.1
npm  3.10.10
node 6.9.5 x64
python 3.6.2
git 2.13.3

$ uname -a
Linux Alex-ArchMainPC 4.11.9-1-ARCH #1 SMP PREEMPT Wed Jul 5 18:23:08 CEST 2017 x86_64 GNU/Linux
```

### Additional Information

Missing.

**Comment by maxbrunsfeld** [comment]
Hi, thanks for pointing out this missing feature; it's not something we've attempted to implement. Currently, when you close a buffer, we don't hold onto any of that buffer's state.

> 4. Restore the last closed tab (Ctrl+Shift+T)
Undo/redo work (not in Safe Mode)

I'm surprised to hear that this works for you; I don't think we have any logic for doing this. Maybe you have some package installed that provides this behavior, but the package does not work when you reopen Atom?

**Comment by AlexWayfer** [comment]
> I'm surprised to hear that this works for you; I don't think we have any logic for doing this. Maybe you have some package installed that provides this behavior, but the package does not work when you reopen Atom?

Oh, OK, I think it's a plugin, sorry (on another computer, can't check now). Thank you, my mistake.

I understood: `Undo`/`Redo` buffer works for Atom re-open, not file.

Can I change this issue for the wish of implementation `Undo`/`Redo` after file re-opening, or should I create a new one?

**Comment by maxbrunsfeld** [comment]
Yeah, I've updated the title of the issue to reflect the feature request. If you want, you could edit the body slightly to explain what you'd want Atom to do, and if there are any other similar text editors that do this.

**Comment by zypA13510** [comment]
Today I deleted some code to commit the change, and I thought I would simply undo that later to continue. Unfortunately, something got in the way. And by the time I realized, I've already closed Atom long ago. Then I reopened Atom, astonished, as I find the undo still works! I don't even know which package (or Atom itself) provides such a feature! It really saved my day.

**Comment by trusktr** [comment]
Vim (and NeoVim) have the ability to do this out of the box.

Not only do they track history the history, but when you undo changes, then make new ones, it branches to a new line of history, so it keeps all branches of history you've ever done to the file.

Vim has key bindings for regular undo/redo (`u` and `ctrl-r` respectively), as well as bindings for traversing all undo/redo history regardless of branch in chronological order (`g-` and `g+`, respectively).

Then there's plugins like `gundo` that can show you a visual of your history, and let you click select items to navigate to.

For example, here's a screenshot showing the history of a file (rendered by `gundo`), including a branch where I undid some changes then started making new changes:

<img width="763" alt="screen shot 2018-11-04 at 2 28 26 pm" src="https://user-images.githubusercontent.com/297678/47970708-f736f100-e03d-11e8-9763-94cb7aa5366e.png">


As you can see there, it shows the changes in chronological order (newest at the top), and also the branches. Pressing undo (`u`) makes it traverse up the tree to the parent in the tree, until you get to the root. Pressing redo (`ctrl-r`) makes it traverse all the way back to the leaf of the branch you were on. Pressing `g-` makes it traverse chronologically in the same order you see in the graph, and opposite for `g+`.

Here's a video of the history tree visible while I make branches of history: 

![history-tree](https://user-images.githubusercontent.com/297678/47970779-e33fbf00-e03e-11e8-9313-5cb06799860a.gif)


I have Vim/NeoVim configured to keep the history cached on the filesystem, so if I restart (Neo)Vim and open the same file, I have all the undo/redo history in place and can continue to undo/redo and view the history.

**Comment by trusktr** [comment]
Note, the green highlight if showing me the diff of what changed in the record, and only visible while the undo tree is open.

**Comment by trusktr** [comment]
I landed on this issue, because due to some tooling, I lost all the work I did over the past two days. This feature would've come to the rescue and been a real time saver!

**Comment by trusktr** [comment]
The [`persistent-undo`](https://atom.io/packages/persistent-undo) package does the trick! Let me see if it works long term with no issues...

The [`undo-tree`](https://atom.io/packages/undo-tree) create a history tree.

`undo-tree` makes a history provider with `buffer.setHistoryProvider`. Does this mean `persistent-undo` will cache it to disk?

Then we need a way to traverse the current branch rather than the whole tree (undo-tree traverse the whole tree).

And then lastly (and optionally) we need is a way to visualize it and click on history items to go there directly.

**Comment by trusktr** [comment]
Yep, verified, `undo-tree` works paired with `persistent-undo`!

However, if you have the `vim-mode-plus` package, then `vim-mode-plus` will override the history that `undo-tree` creates, so `undo-tree` doesn't work in this case. I opened an issue in `vim-mode-plus`: https://github.com/t9md/atom-vim-mode-plus/issues/1101.

---

Instead of having plugins implement undo history (f.e. `undo-tree` and `vim-mode-plus` both doing their own thing), seems like the tree functionality needs to be centralized in Atom itself, then `undo-tree` would not be needed, `persistent-undo` would still be useful, and `vim-mode-plus` would simply continue to provide the chunks that the user inputted (f.e. while in insert mode).

**Comment by AlexWayfer** [comment]
> Verify that you can still reproduce the issue in the latest version of Atom

Yes.

> What version of Atom you reproduced the issue on

`1.41.0`

> What OS and version you reproduced the issue on

Arch Linux.

> What steps you followed to reproduce the issue

The same as in the issue description.

**Comment by AlexWayfer** [comment]
Atom 1.53.0, no progression of this issue.
